### PR TITLE
When redacting, keep event fields around that maintain the historical event structure intact (MSC2716)

### PR DESCRIPTION
**Pull request by MadLittleMods**
When redacting, keep event fields around that maintain the historical event structure intact.

Fix https://github.com/matrix-org/synapse/issues/10521

Part of MSC2716: https://github.com/matrix-org/matrix-doc/pull/2716

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [ ] ~~Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)~~
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

**Comment by erikjohnston** [comment]
This looks good, though does mean that existing events will end up with different event IDs. I think we should probably update the room version ID to avoid mass confusion?

**Comment by MadLittleMods** [comment]
@erikjohnston I'm not following 😣. Does redacting an event give it a different event ID?

These redaction checks are already under the experimental MSC2716 room version? Do we need to bump that version again?

**Comment by erikjohnston** [comment]
> @erikjohnston I'm not following . Does redacting an event give it a different event ID?
> 
> These redaction checks are already under the experimental MSC2716 room version? Do we need to bump that version again?

Ah, sorry, its non obvious how this relates to event IDs. In non-v1 room versions event IDs are the [reference hashes](https://spec.matrix.org/unstable/server-server-api/#calculating-the-reference-hash-for-an-event) for the event, which are calculated by redacting the event, serialising it to canonical JSON and then calculating the SHA256 hash of those bytes. Hence, if you change the redaction algorithm then the canonical JSON bytes will be different and therefore so would the hash.<sup>1</sup>.



> These redaction checks are already under the experimental MSC2716 room version? Do we need to bump that version again?

I would, just to prevent confusion with any existing test rooms using that experimental version. I hope that's fine and easy?





---
<sup>1</sup> Note that an event also contains a [content hash](https://spec.matrix.org/unstable/server-server-api/#calculating-the-content-hash-for-an-event), which is a hash over the *un*redacted event and is included in the redacted event. This is what protects the unredacted event fields from being changed.

**Comment by MadLittleMods** [comment]
> Ah, sorry, its non obvious how this relates to event IDs. In non-v1 room versions event IDs are the [reference hashes](https://spec.matrix.org/unstable/server-server-api/#calculating-the-reference-hash-for-an-event) for the event, which are calculated by redacting the event, serialising it to canonical JSON and then calculating the SHA256 hash of those bytes. Hence, if you change the redaction algorithm then the canonical JSON bytes will be different and therefore so would the hash1.

Thanks for that context! That makes sense 👍 

---

Do we actually verify the `event_id` hashes after it's created? I assume this only matters for remote federated homeservers trying to verify a received event?

Just as a thought, if redacting an event messes up the event verification, that essentially `soft_fails` the event which has the same effect as redacting 🤔 (not visible for client)

I understand why it matters, just asking a "technically, does it work? 🤓" kind of question 😁

**Comment by erikjohnston** [comment]
> Do we actually verify the `event_id` hashes after it's created? I assume this only matters for remote federated homeservers trying to verify a received event that was redacted?

Amusingly, yes, as the Event object calculates the `event_id` on creation, rather than reading it from the DB. The `get_events(..)` function returns the event as part of an `event.event_id -> event` mapping, so if you ask for event `$A` you'd get back a dict with `$B` in instead, which'd then cause the code to assume the server didn't have `$A`

> Just as a thought, if redacting an event messes up the event verification, that essentially `soft_fails` the event which has the same effect as redacting  (not visible for client)
> 
> I understand why it matters, just asking a "technically, does it work? " kind of question

It will reject the event rather than "soft fail", but yes it'd have the same impact as redacting the event. I think not changing the redaction rules would mostly be fine in practice (we have in the past forgotten to change the redaction algorithm when we've changed the auth rules).

**Review by erikjohnston — APPROVED** [review]

File: changelog.d/10538.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+When redacting, keep event fields around that maintain the MSC2716 historical event structure intact.
```
Comment (erikjohnston) [inline changelog.d/10538.feature:1]:
```suggestion
Add support for new redaction rules for historical events specified in [MSC2716](https://github.com/matrix-org/matrix-doc/pull/2716).
```

**Comment by MadLittleMods** [comment]
> Amusingly, yes, as the Event object calculates the `event_id` on creation, rather than reading it from the DB. The `get_events(..)` function returns the event as part of an `event.event_id -> event` mapping, so if you ask for event `$A` you'd get back a dict with `$B` in instead, which'd then cause the code to assume the server didn't have `$A`

😁 That would be an interesting scenario

---

Thanks for the review @erikjohnston 🐘

**Pull request merged**